File: scripts/benchmark/summarize_arc_v3_campaign.py
"""Preserve paired compiler-campaign evidence; large raw studies remain local."""
import argparse, gzip, hashlib, json
from pathlib import Path


parser = argparse.ArgumentParser()
parser.add_argument('--name', required=True)
parser.add_argument('--baseline', required=True)
parser.add_argument('--target', required=True, type=float)
args = parser.parse_args()
assert args.name.startswith('arc-') and args.name[4:].isdigit()
assert 0 < args.target < 1000
root = Path('generated/benchmark-v3') / args.name


def read(path):
    body = path.read_bytes()
    digest = hashlib.sha256(body).hexdigest()
    assert digest == Path(str(path) + '.sha256').read_text().strip(), path
    return json.loads(gzip.decompress(body) if path.suffix == '.gz' else body), digest


baseline_path = Path(args.baseline)
baseline, baseline_sha = read(baseline_path)
base = {r['sourceId']: r for r in baseline['rows'] if r['seed'] == 16}
assert len(base) == 88
studies = []
for path in sorted(root.glob('*/run.json.gz')):
    run, digest = read(path)
    rows, plan = run['rows'], run['plan']
    assert plan['budget'] == 750000 and plan['seed'] == 16
    assert len(rows) == len({r['sourceId'] for r in rows})
    assert set(r['sourceId'] for r in rows) == set(plan['sources'])
    paired = []
    for r in rows:
        b = base[r['sourceId']]
        frames = r['resources']['physicalFrames']
        assert 0 <= frames <= plan['budget']
        paired.append(dict(id=r['sourceId'], before=b['score']['score'], after=r['score']['score'],
                           delta=r['score']['score']-b['score']['score'], valid=r['score']['valid'],
                           physicalFrames=frames, trackHash=r['trackHash'],
                           sameTrack=r['trackHash']==b['trackHash'],
                           sameFrames=frames==b['resources']['physicalFrames']))
    full = set(r['sourceId'] for r in rows) == set(base)
    assert full == ('headline' in run['summary'])
    patch_path=path.parent/'compiler.patch'
    patch=patch_path.read_bytes()
    patch_sha=hashlib.sha256(patch).hexdigest()
    preserved_patch=None
    if patch and len(patch)<=100000:
        preserved_patch=Path('benchmark/v3/studies')/(args.name+'-prototypes')/(patch_sha+'.patch')
        preserved_patch.parent.mkdir(parents=True,exist_ok=True)
        preserved_patch.write_bytes(patch)
        Path(str(preserved_patch)+'.sha256').write_text(patch_sha+'\n')
    studies.append(dict(name=path.parent.name,path=str(path),sha256=digest,
                        compiler=plan['compiler']['candidateFingerprint'],head=plan['compiler']['head'],
                        options=plan['options'],modelSha256=plan.get('modelSha256'),
                        valueModelSha256=plan.get('valueModelSha256'),cases=len(rows),fullSuite=full,
                        headline=run['summary'].get('headline'),valid=sum(r['valid'] for r in paired),
                        arithmeticMeanDelta=sum(r['delta'] for r in paired)/len(paired),
                        improved=sum(r['delta']>0 for r in paired),regressed=sum(r['delta']<0 for r in paired),
                        physicalFrames=sum(r['physicalFrames'] for r in paired),paired=paired,
                        compilerPatch=dict(path=str(patch_path),sha256=patch_sha,bytes=len(patch),
                                           preservedPath=str(preserved_patch) if preserved_patch else None)))
canonical_path=Path(f'benchmark/v3/studies/{args.name}-validation.json')
canonical=None
if canonical_path.exists():
    validation,digest=read(canonical_path)
    assert validation['goal']==args.target and validation['targetReached']
    canonical=dict(path=str(canonical_path),sha256=digest,headline=validation['summary']['headline'],
                   compilerCommit=validation['compilerCommit'])
unfinished=[]
for path in sorted(root.glob('full-*/plan.json')):
    if (path.parent/'run.json.gz').exists():continue
    plan,digest=read(path)
    unfinished.append(dict(name=path.parent.name,planSha256=digest,options=plan['options'],
                           completedCells=len(list(path.parent.glob('*.json.gz'))),
                           note='No complete result or headline; this plan is pending, running or was stopped before a full panel was completed.'))
record = dict(schema='line.arc-v3-campaign-research.v1',status='complete' if canonical else 'in-progress',target=args.target,
              baseline=dict(path=str(baseline_path),sha256=baseline_sha,headline=baseline['summary']['headline']),
              targetComparison='strictly-greater-than',canonical=canonical,
              note='Exposed development research. Partial pilots have no headline. Full single-seed results require canonical confirmation on seeds 16/17 through the public compiler. Failed runs retain zero scores.',
              studies=studies,unfinished=unfinished)
out=Path(f'benchmark/v3/studies/{args.name}-research.json')
body=(json.dumps(record,indent=2,allow_nan=False)+'\n').encode()
out.write_bytes(body);Path(str(out)+'.sha256').write_text(hashlib.sha256(body).hexdigest()+'\n')
print(json.dumps([dict(name=s['name'],headline=s['headline'],meanDelta=s['arithmeticMeanDelta'],valid=s['valid'],cases=s['cases']) for s in studies],indent=2))

File: scripts/benchmark/summarize_arc_v3_campaign.ts
/** Compact successful and adverse V3 campaign evidence; raw studies stay local. */
import assert from 'node:assert/strict';
import {readFileSync,writeFileSync,readdirSync,existsSync,mkdirSync} from 'node:fs';
import {gunzipSync} from 'node:zlib';
import {loadCases,sha} from '../../benchmark/v3/model.ts';
import {summarize} from '../../benchmark/v3/evaluator.ts';
const root='generated/benchmark-v3/arc-900';
const read=(p:string)=>{const b=readFileSync(p);assert.equal(sha(b),readFileSync(p+'.sha256','utf8').trim());return JSON.parse((p.endsWith('.gz')?gunzipSync(b):b).toString());};
const cases=loadCases(),baseline=read('benchmark/v3/runs/initial-after.json.gz');
const base=new Map(baseline.rows.filter((r:any)=>r.seed===16).map((r:any)=>[r.sourceId,r]));
const studies=readdirSync(root).sort().flatMap(name=>{
  const path=`${root}/${name}/run.json.gz`;if(!existsSync(path))return[];
  const run=read(path);assert.equal(run.plan.seed,16);assert.equal(run.plan.budget,750000);
  const full=run.rows.length===cases.length,summary=full?summarize(run.rows,cases,[16]):null;
  if(full)assert.deepEqual(summary,run.summary);
  const paired=run.rows.map((r:any)=>({id:r.sourceId,score:r.score.score,delta:r.score.score-(base.get(r.sourceId) as any).score.score,valid:r.score.valid,
    physicalFrames:r.stats.sim_frames,trackHash:r.trackHash}));
  assert.ok(paired.every((r:any)=>r.physicalFrames<=750000));
  return [{name,path,sha256:sha(readFileSync(path)),compiler:run.plan.compiler.candidateFingerprint,head:run.plan.compiler.head,
    options:run.plan.options,modelSha256:run.plan.modelSha256??null,valueModelSha256:run.plan.valueModelSha256??null,cases:paired.length,valid:paired.filter((r:any)=>r.valid).length,
    headline:summary?.headline??null,strata:summary?.strata??null,
    arithmeticMean:paired.reduce((s:number,r:any)=>s+r.score,0)/paired.length,
    improved:paired.filter((r:any)=>r.delta>0).length,regressed:paired.filter((r:any)=>r.delta<0).length,
    totalPhysicalFrames:paired.reduce((s:number,r:any)=>s+r.physicalFrames,0),paired,
    finalizationRecovery:run.finalizationRecovery??null}];
});
const record={schema:'line.arc-v3-900-research.v1',status:'active',target:900,
  baseline:{headline:baseline.summary.headline,commit:baseline.plan.compiler.commit,seeds:baseline.plan.seeds,suiteFingerprint:baseline.plan.suiteFingerprint},
  note:'These are exposed development studies. Single-seed full-suite readings require canonical two-seed confirmation through the public compiler. Partial panels have no headline. Failed physical runs retain zero scores. Search/model choices must be tested together, not added arithmetically.',
  bestFullSuiteResearch:Math.max(...studies.flatMap(s=>s.headline===null?[]:[s.headline])),studies,
  constraints:['Frozen V3 inputs, scorer, detector, physics and actual-frame allowance.','Substantial coherent normal arcs; no acceleration or isolated controls.','No new owner audiovisual approval.','Large raw studies and models remain local until a selected compiler artifact is published.']};
mkdirSync('benchmark/v3/studies',{recursive:true});
const body=JSON.stringify(record,null,2)+'\n',out='benchmark/v3/studies/arc-900-research.json';writeFileSync(out,body);writeFileSync(out+'.sha256',sha(body)+'\n');
console.log(JSON.stringify({studies:studies.length,best:record.bestFullSuiteResearch,out}));
